src/v/cluster/data_migration_table.cc

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "cluster/data_migration_table.h"
+
+#include "cluster/controller_snapshot.h"
+#include "cluster/data_migrated_resources.h"
+#include "cluster/data_migration_types.h"
+#include "cluster/errc.h"
+#include "cluster/logger.h"
+#include "cluster/topic_table.h"
+
+#include <seastar/util/variant_utils.hh>
+
+namespace cluster::data_migrations {
+
+migrations_table::migrations_table(
+  ss::sharded<migrated_resources>& resources, ss::sharded<topic_table>& topics)
+  : _resources(resources)
+  , _topics(topics) {}
+
+bool migrations_table::is_valid_state_transition(state current, state target) {
+    switch (current) {
+    case state::planned:
+        return target == state::preparing;
+    case state::preparing:
+        return target == state::prepared || target == state::canceling;
+    case state::prepared:
+        return target == state::executing || target == state::canceling;
+    case state::executing:
+        return target == state::executed || target == state::canceling;
+    case state::executed:
+        return target == state::finished || target == state::canceling;
+    case state::canceling:
+        return target == state::cancelled;
+    /**
+     * Those are the terminal states, there it is impossible to get out of them
+     * in other way than deleting migration object
+     **/
+    case state::cancelled:
+        [[fallthrough]];
+    case state::finished:
+        return false;
+    }
+}
+
+bool migrations_table::is_empty_migration(const data_migration& m) {
+    return ss::visit(
+      m, [](const auto& m) { return m.topics.empty() && m.groups.empty(); });
+}
+
+ss::future<std::error_code>
+migrations_table::apply_update(model::record_batch batch) {
+    auto cmd = co_await deserialize(std::move(batch), commands);
+
+    co_return co_await std::visit(
+      [this](auto cmd) { return apply(std::move(cmd)); }, std::move(cmd));
+}
+
+ss::future<>
+migrations_table::fill_snapshot(controller_snapshot& snapshot) const {
+    snapshot.data_migrations.next_id = _next_id;
+    snapshot.data_migrations.migrations.reserve(_migrations.size());
+    for (auto& [id, migration] : _migrations) {
+        snapshot.data_migrations.migrations.emplace(id, migration.copy());
+    }
+
+    co_return;
+}
+
+ss::future<> migrations_table::apply_snapshot(
+  model::offset, const controller_snapshot& snapshot) {
+    _next_id = snapshot.data_migrations.next_id;
+    _migrations.reserve(_migrations.size());
```

Comment:
:rofl:

Reply:
This function also does not account for removing entries, I'll be fixing it in my PR so no action needed here.